ts/src/header-validator/maybe.ts

```diff
@@ -1,16 +1,10 @@
-export type Maybeable<T> = T | Maybe<T>
-
 export class Maybe<T> {
   static readonly None = new Maybe<never>()
 
   static some<T>(this: void, t: T): Maybe<T> {
     return new Maybe(t)
   }
 
-  static flatten<T>(this: void, t: Maybeable<T>): Maybe<T> {
```

Comment:
Just to make sure I understand correctly.

The edge case is that it doesn't support nested Maybe types. 

If T is a Maybe<number>, it wont return Maybe<Maybe<number>> as expected by the type and will instead return a Maybe<number>. Correct ?

Do we have a use case for these ? 

Intuitively, I would be using .flatten to avoid that case and ensure that I always get back to a Maybe<T>.

Reply:
> If T is a `Maybe<number>`, it wont return `Maybe<Maybe<number>>` as expected by the type and will instead return a Maybe. Correct ?

Correct.

I think it is possible to prevent this using conditional types with `never`, but then all code that transitively calls `flatten` needs to propagate those bounds as well. So I do think it's simpler to just have separate `map` and `flatMap`.

Reply:
Yeah, I'm not sure it's possible to statically type it and prevent incorrect inference.